lib/msal-browser/src/cache/TokenCache.ts

```diff
@@ -0,0 +1,120 @@
+
+export class TokenCache implements ITokenCache {
+    protected config: BrowserConfiguration;
+    public isBrowserEnvironment: boolean;
+    private storage: BrowserCacheManager;
+    private logger: Logger;
+    private cryptoObj: ICrypto;
+
+    constructor(configuration: Configuration, storage: BrowserCacheManager, logger: Logger, cryptoObj: ICrypto) {
+        this.isBrowserEnvironment = typeof window !== "undefined";
+        this.config = buildConfiguration(configuration, this.isBrowserEnvironment);
+        this.storage = storage;
+        this.logger = logger;
+        this.cryptoObj = cryptoObj;
+    }
+    
+    // Move getAllAccounts here and cache utility APIs
+
+    loadTokens(request: SilentRequest, response: ServerAuthorizationROPCResponse, options: LoadTokenOptions): void {
+        this.logger.info("TokenCache - loadTokens called");
+
+        if (!response.id_token) {
+            throw BrowserAuthError.createUnableToLoadTokenError("Please ensure server response includes id token.");
+        }
+
+        if (request.account) {
+            this.loadIdToken(response.id_token, request.account.homeAccountId, request.account.environment, request.account.tenantId, options);
+
+            this.loadAccessToken(request, response, options, request.account.homeAccountId, request.account.environment, request.account.tenantId);
+        } else if (request.authority && options.clientInfo) {
+            const homeAccountId = `${options.clientInfo?.uid}.${options.clientInfo?.utid}`;
+
+            const authorityOptions: AuthorityOptions = {
+                protocolMode: this.config.auth.protocolMode,
+                knownAuthorities: this.config.auth.knownAuthorities,
+                cloudDiscoveryMetadata: this.config.auth.cloudDiscoveryMetadata,
+                authorityMetadata: this.config.auth.authorityMetadata
+            };
+            const authority = new Authority(request.authority, this.config.system.networkClient, this.storage, authorityOptions);
+
+            this.loadIdToken(response.id_token, homeAccountId, authority.hostnameAndPort, authority.tenant, options);
+
+            this.loadAccessToken(request, response, options, homeAccountId, authority.hostnameAndPort, authority.tenant);
+
+        } else {
+            throw BrowserAuthError.createUnableToLoadTokenError("Please provide a request with an account, or a request with authority and clientInfo.");
+        }
+    }
+
+    private loadIdToken(idToken: string, homeAccountId: string, environment: string, tenantId: string, options: LoadTokenOptions): void {
+
+        const idTokenEntity = IdTokenEntity.createIdTokenEntity(homeAccountId, environment, idToken, this.config.auth.clientId, tenantId);
+
+        if (this.isBrowserEnvironment) {
+            this.logger.verbose("TokenCache - loading id token");
+            this.storage.setIdTokenCredential(idTokenEntity);
+        } else {
+            if (options.callback) {
```

Comment:
Should a provided callback be given precedence over default window storage (this.isBrowserEnvironment condition)?

Reply:
In our design discussions, we decided to store in local/session storage by default, but to provide a way to use a callback if it wasn't available. Is there a reason we'd want the callback to take precedence?

Reply:
I don't think we should make assumptions about how people would use this API. If we offer a callback override we should honor it. It feels weird to offer an option that would be totally ignored in some scenarios.

Reply:
That's fair. Updated.